distributed/worker.py

```diff
@@ -3445,47 +3461,78 @@ async def find_missing(self) -> None:
-        try:
-            for dep, workers in who_has.items():
-                if not workers:
-                    continue
+    def _update_who_has(
+        self, who_has: dict[str, Collection[str]], *, stimulus_id: str
+    ) -> RecsInstrs:
+        recs: Recs = {}
+        instructions: Instructions = []
 
-                if dep in self.tasks:
-                    dep_ts = self.tasks[dep]
-                    if self.address in workers and self.tasks[dep].state != "memory":
-                        logger.debug(
-                            "Scheduler claims worker %s holds data for task %s which is not true.",
-                            self.name,
-                            dep,
+        for key, workers in who_has.items():
+            ts = self.tasks.get(key)
+            if not ts:
+                continue
+            workers = set(workers)
+
+            if self.address in workers:
+                workers.remove(self.address)
+                if ts.state != "memory":
+                    logger.debug(
+                        "Scheduler claims worker %s holds data for task %s, "
+                        "which is not true.",
+                        self.name,
+                        ts,
+                    )
+                    instructions.append(
+                        MissingDataMsg(
+                            key=key, errant_worker=self.address, stimulus_id=stimulus_id
                         )
-                        # Do not mutate the input dict. That's rude
-                        workers = set(workers) - {self.address}
-                    dep_ts.who_has.update(workers)
+                    )
 
-                    for worker in workers:
-                        self.has_what[worker].add(dep)
-                        self.data_needed_per_worker[worker].push(dep_ts)
-        except Exception as e:  # pragma: no cover
-            logger.exception(e)
-            if LOG_PDB:
-                import pdb
+            if ts.who_has == workers:
+                continue
 
-                pdb.set_trace()
-            raise
+            self.log.append(
+                (
+                    key,
+                    "update-who-has",
+                    list(ts.who_has),
+                    list(workers),
+                    stimulus_id,
+                    time(),
+                )
+            )
+
+            for worker in ts.who_has - workers:
+                self.has_what[worker].discard(key)
+                # Can't remove from self.data_needed_per_worker; there is logic
+                # in _select_keys_for_gather to deal with this
+
+            for worker in workers - ts.who_has:
+                self.has_what[worker].add(key)
+                if ts.state == "fetch":
+                    self.data_needed_per_worker[worker].push(ts)
+
+            ts.who_has = workers
+            if not workers and ts.state == "fetch":
+                recs[ts] = "missing"
```

Comment:
I've been touching the update_who_has a bit too often in combination with the state machine. I started from a point without recommendations, added recommendations, removed them again, and now we're adding them back in. 
My gut tells me we should take a step back and reevaluate what we're actually doing here and why. The transition system was not intended to make state modifications like these more complicated but the opposite. For instance, I am concerned that only checking for fetching tasks is not sufficient.

Before we "overhaul" this mechanism yet again we should talk about what this mechanism is actually supposed to do and how it connects to other mechanisms.

Reply:
Should we have a call?
In advance of it, please look at the next PR in the iteration, #6348, which adds further transition logic to update_who_has.

Reply:
I'm reviewing what happened while I was gone and will contact you once I'm up to speed

Reply:
Could you come up with a case that won't work?